adding an atime test case  for gshift as gforce optimized shift  #5205

DorisAmoakohene opened the pull request:
[In this test case the user mentions that the shift() function in data.table version 1.9.6 is slow when applied to many groups](https://github.com/Rdatatable/data.table/issues/1534). The exact cause or specific use case is not elaborated upon in the issue.[visit here to read more](https://github.com/Rdatatable/data.table/issues/1534)

[The pull request address the issue by allowing GForce optimization for lapply even without using .SD1. ](https://github.com/Rdatatable/data.table/pull/5205)

tdhock commented:
Hi I converted to draft because I think you should first merge your first PR, and then use what you learn from that experience to improve your other PRs, before asking others to review them.

DorisAmoakohene commented:
> Hi I converted to draft because I think you should first merge your first PR, and then use what you learn from that experience to improve your other PRs, before asking others to review them.

I have made most corrections to this PR  for review and merge

tdhock commented:
please resolve conflicts and fix typo shif -> shift

DorisAmoakohene commented:
> please resolve conflicts and fix typo shif -> shift

I have resolved this. Thank you

tdhock commented on .ci/atime/tests.R:
```diff
@@ -134,4 +135,42 @@ test.list <- atime::atime_test_list(
+    Slower = "cacdc92df71b777369a217b6c902c687cf35a70d"), # Circa 2020 parent of the first commit (https://github.com/Rdatatable/data.table/commit/74636333d7da965a11dad04c322c752a409db098) in the PR (https://github.com/Rdatatable/data.table/pull/4501/commits) that fixes the issue 
+
+  NULL)
+>>>>>>> master
```
you need to remove >>> and === which are git annotations, not R code

tdhock commented on .ci/atime/tests.R:
```diff
@@ -171,6 +170,4 @@ test.list <- atime::atime_test_list(
     Slow = "3ca83738d70d5597d9e168077f3768e32569c790", # Circa 2024 master parent of close-to-last merge commit (https://github.com/Rdatatable/data.table/commit/353dc7a6b66563b61e44b2fa0d7b73a0f97ca461) in the PR (https://github.com/Rdatatable/data.table/pull/4501/commits) that fixes the issue 
     Slower = "cacdc92df71b777369a217b6c902c687cf35a70d"), # Circa 2020 parent of the first commit (https://github.com/Rdatatable/data.table/commit/74636333d7da965a11dad04c322c752a409db098) in the PR (https://github.com/Rdatatable/data.table/pull/4501/commits) that fixes the issue 
 
-  NULL)
```
please undo deletion of this line which is required to terminate the test list

Anirban166 reviewed (approved):
Looks good to go!

tdhock commented:
currently we see this result 
![image](https://github.com/user-attachments/assets/8d53a254-8e4f-4bbd-8a19-31e7540a3434)
there are two issues
1. there is not a huge separation between Regression and Fixed at largest N, but there is some separation at smaller N, why? Can you please investigate, perhaps with a larger N? and double check the historical commit IDs?
2. There is no Before, so please either add that, or change to Fast and Slow.

DorisAmoakohene commented:
> currently we see this result ![image](https://private-user-images.githubusercontent.com/932850/367509261-cedfea83-817a-4109-824b-92c656040cc9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY1MTU2MjUsIm5iZiI6MTcyNjUxNTMyNSwicGF0aCI6Ii85MzI4NTAvMzY3NTA5MjYxLWNlZGZlYTgzLTgxN2EtNDEwOS04MjRiLTkyYzY1NjA0MGNjOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxNlQxOTM1MjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05YTM1NWZlYzk1Zjk5N2E4YzEzMmE4NWRiNmYwNzM1ZWU3NWU4ZGQwZmYyZjM3NzdmOTFhZDYzN2ZhYjZkOTA4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9._0gsd3JT1dK0hpktkDpcS1BB3oy2mSAKUQpnTWAFQYM)
> 
> there are three issues
> 
> 1. there is not a huge time separation between Regression and Fixed at largest N, but there is some separation at smaller N, why? Can you please investigate, perhaps with a larger N? and double check the historical commit IDs?
> 2. There is no Before, so please either add that, or change to Fast and Slow.
> 3. why does Regression use less memory than Fixed?

Alright

DorisAmoakohene commented:
> currently we see this result ![image](https://private-user-images.githubusercontent.com/932850/367509261-cedfea83-817a-4109-824b-92c656040cc9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY1MTg1NzIsIm5iZiI6MTcyNjUxODI3MiwicGF0aCI6Ii85MzI4NTAvMzY3NTA5MjYxLWNlZGZlYTgzLTgxN2EtNDEwOS04MjRiLTkyYzY1NjA0MGNjOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxNlQyMDI0MzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xMDkzYmYzMTMwOTdhMzQzNDZlNzVjY2U5NTUyMWNlMjY2MjNkZGY3ZDA3NzAzMTVmMWRkYWJlZmM3Y2E4ZDM4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.sGeNaowIxydLM6Twnq3DqOz8lSIBQuEpdfMI7WlWX6c)
> 
> there are three issues
> 
> 1. there is not a huge time separation between Regression and Fixed at largest N, but there is some separation at smaller N, why? Can you please investigate, perhaps with a larger N? and double check the historical commit IDs?
> 2. There is no Before, so please either add that, or change to Fast and Slow.
> 3. why does Regression use less memory than Fixed?

I investigated further with a larger sample size (N), but found no difference, it stops when it gets to 10000

. Even when I reverted to the  commit to the commit before the last commits, where the PR was merge nothing changes.

<img width="510" alt="pic5205" src="https://github.com/user-attachments/assets/c42eed17-3adc-4fd4-9e7d-d05c5697c9a5">

tdhock commented:
the original benchmarks were reported by @ben-schwen in https://github.com/Rdatatable/data.table/pull/5205#issuecomment-943509212
there are two sizes in the benchmarks. 
* number of rows of DT
* number of values in sample (y column)

your current setup has ` DT = data.table(x = sample(N), y = sample(1e2, N, TRUE))`  so N controls number of rows of DT, not number of values in sample (y column) which is currently fixed at 1e2. Maybe you could try varying that instead?
Or varying both?

MichaelChirico commented:
@tdhock what's the status here?

tdhock commented:
let's close for now and open another pr with a similar test case.
